packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,226 @@
+        componentVMs = new Set();
+        activeComponents.set(Ctor, componentVMs);
+    }
+    // this will allow us to keep track of the hot components
+    componentVMs.add(vm);
+
+    // tracking active template
+    const tpl = vm.cmpTemplate;
+    let templateVMs = activeTemplates.get(tpl);
+    if (isUndefined(templateVMs)) {
+        templateVMs = new Set();
+        activeTemplates.set(tpl, templateVMs);
+    }
+    // this will allow us to keep track of the templates that are
+    // being used by a hot component
+    templateVMs.add(vm);
+
+    // tracking active styles associated to template
+    const stylesheets = tpl.stylesheets;
+    if (!isUndefined(stylesheets)) {
+        stylesheets.forEach((stylesheet) => {
+            let stylesheetVMs = activeStyles.get(stylesheet);
+            if (isUndefined(stylesheetVMs)) {
+                stylesheetVMs = new Set();
+                activeStyles.set(stylesheet, stylesheetVMs);
+            }
+            // this will allow us to keep track of the stylesheet that are
+            // being used by a hot component
+            stylesheetVMs.add(vm);
+        });
+    }
+}
+
+export function removeActiveVM(vm: VM) {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    // tracking inactive component
+    const Ctor = vm.def.ctor;
+    let list = activeComponents.get(Ctor);
+    if (!isUndefined(list)) {
+        // deleting the vm from the set to avoid leaking memory
+        list.delete(vm);
+    }
+    // removing inactive template
+    const tpl = vm.cmpTemplate;
+    list = activeTemplates.get(tpl);
+    if (!isUndefined(list)) {
+        // deleting the vm from the set to avoid leaking memory
+        list.delete(vm);
+    }
+    // removing active styles associated to template
+    const styles = tpl.stylesheets;
+    if (!isUndefined(styles)) {
+        styles.forEach((style) => {
+            list = activeStyles.get(style);
+            if (!isUndefined(list)) {
+                // deleting the vm from the set to avoid leaking memory
+                list.delete(vm);
+            }
+        });
+    }
+}
+
+export function swapTemplate(oldTpl: Template, newTpl: Template): boolean {
+    if (process.env.NODE_ENV !== 'production') {
+        swappedTemplateMap.set(oldTpl, newTpl);
+        return rehydrateHotTemplate(oldTpl);
+    }
+    return false;
+}
+
+export function swapComponent(
+    oldComponent: ComponentConstructor,
+    newComponent: ComponentConstructor
+): boolean {
+    if (process.env.NODE_ENV !== 'production') {
+        swappedComponentMap.set(oldComponent, newComponent);
+        return rehydrateHotComponent(oldComponent);
+    }
+    return false;
+}
+
+export function swapStyle(oldStyle: StylesheetFactory, newStyle: StylesheetFactory): boolean {
+    if (process.env.NODE_ENV !== 'production') {
+        swappedStyleMap.set(oldStyle, newStyle);
```

Comment:
I don't think we have a way to validate whether a style is a compiled style or not... probably we should eventually do some registration from compiler to make sure that people don't mess around with this.